api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,600 @@
+	if len(incompatibilityErrs) > 0 {
+		multiError = errors.New(errors.IncompatibleConsumerAndProvider, errors.MultiError(incompatibilityErrs))
+	}
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	return resourceProviders, unusedProviders, desiredScale, multiError
+}
+
+func (e *EnvironmentScaler) ScaleToState(clusterName string, desiredScale int, unusedProviders []*scaler.ResourceProvider) error {
+	asg, err := readASG(e.Client.AutoScaling, clusterName)
+	if err != nil {
+		return err
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+
+	log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+
+	asgName := aws.StringValue(asg.AutoScalingGroupName)
+	desiredScale64 := int64(desiredScale)
+
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+
+	// sample error returned when desired < min || desired > max:
+	//
+	// 2017/12/14 23:04:39 [ERROR] [Worker] [3]: Failed to run job 1513321478660496503: ValidationError: Desired capacity:1 must be between the specified min size:2 and max size:2
+	//         status code: 400, request id: ed313128-e165-11e7-ab64-c9c2b806a2ed
+	if err := updateASG(e.Client.AutoScaling, asgName, nil, nil, &desiredScale64); err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "ValidationError" {
+			text := fmt.Sprintf("[INFO] [EnvironmentScaler] Tried to scale above/below max/min scale for environment '%s'. ", clusterName)
+			text += fmt.Sprintf("(Desired: %d, Min: %d, Max: %d)", desiredScale, minCapacity, maxCapacity)
+			log.Print(text)
+		} else {
+			return err
+		}
+	}
+
+	// choose which instances to terminate during our scale-down process
+	// instead of having asg randomly select instances
+	// e.g. if we scale from 5 to 3, we can terminate up to 2 unused instances
+	if currentCapacity > desiredScale {
+		maxNumberOfInstancesToTerminate := currentCapacity - desiredScale
+
+		canTerminate := func(i int) bool {
+			if i+1 > maxNumberOfInstancesToTerminate {
+				return false
+			}
+
+			if i > len(unusedProviders)-1 {
+				return false
+			}
+
+			return true
+		}
+
+		for i := 0; canTerminate(i); i++ {
+			unusedProvider := unusedProviders[i]
+			instanceID := unusedProvider.ID
+
+			log.Printf("[DEBUG] [EnvironmentScaler] Terminating unused instance '%s' from environment '%s'.", instanceID, clusterName)
+
+			if err := e.terminateInstanceInAutoScalingGroup(instanceID); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*scaler.ResourceProvider, error) {
+	var environment *models.Environment
+
+	val, ok := e.environmentCache.Getf(clusterName)
+
+	switch ok {
+	case true:
+		environment = val.(*models.Environment)
```

Comment:
not having generics 😞

Reply:
doing a `switch` over a `bool` is a little awkward.
This is one of those rare cases where I think an `if else` is more readable. 
```
var environment *models.Environment
v, ok := e.environmentCache.Getf(clusterName)
if ok {
    environment := v.(*models.Environment)
} else {
    ///
}
```